contracts/staking-vault/StakingVault.sol

```diff
@@ -0,0 +1,284 @@
+pragma solidity ^0.8.0;
+
+import "@openzeppelin-4.8.0/token/ERC1155/ERC1155.sol";
+import "@openzeppelin-4.8.0/token/ERC20/utils/SafeERC20.sol";
+import "@openzeppelin-4.8.0/token/ERC20/extensions/IERC20Metadata.sol";
+
+import "../../interfaces/IPriceFeeds.sol";
+import "../proxies/0_8/Upgradeable_0_8.sol";
+
+contract StakingVault is Upgradeable_0_8, ERC1155 {
+  using SafeERC20 for IERC20;
+  struct TokenComposition {
+    address depositToken;
+    address tokenToBack;
+  }
+  address public protocol;
+  IPriceFeeds public priceFeed;
+  address public valuationToken;
+  address[] public stakingTokens;
+  mapping(address => bool) public tokenSupported;
+  mapping(uint256 => uint256) public sTokenPrice;
+  mapping(uint256 => TokenComposition) public identifierToTokens;
+  mapping(uint256 => uint256) public balanceStakedPerID;
+  mapping(uint256 => uint256) public supplyPerID;
+
+  mapping(address => uint256) public undistributedRewards;
+  mapping(uint256 => uint256) public rewardsPerToken;
+  address public rewardToken;
+  mapping(address => mapping(uint256 => uint256)) public lastClaimRewardAccrual;
+  mapping(uint256 => bool) public initialized;
+  mapping(address => uint8) public tokenDecimalCache;
+
+  event PriceFeed(address feed);
+  event Protocol(address protocol);
+  event ValuationToken(address token);
+  event TokensSupported(address[] supportedTokens);
+  event Deposit(address depositor, address tokenBacked, address depositToken, uint256 amount);
+  event Withdraw(address withdrawer, address tokenBacked, address receivedToken, uint256 amount);
+  event RewardToken(address token);
+  event PoolDraw(address tokenBacked, address tokenCovered, uint256 amountCovered);
+  event AddRewards(address tokenBacked, uint256 amount);
+  event AccumulateRewards(address tokenBacked, uint256 amount);
+  event DistributeRewards(address tokenBacked, uint256 amount);
+
+  modifier onlyProtocol() {
+    require(msg.sender == protocol, "not protocol");
+    _;
+  }
+
+  constructor(string memory uri_) ERC1155(uri_) {}
+
+  function updateTokenSupport(
+    address[] memory tokens,
+    bool[] memory support,
+    address[] memory listOfSupportedTokens
+  ) external onlyOwner {
+    stakingTokens = listOfSupportedTokens;
+    for (uint256 i; i < tokens.length; ) {
+      tokenSupported[tokens[i]] = support[i];
+      if (tokenDecimalCache[tokens[i]] == 0) {
+        tokenDecimalCache[tokens[i]] = IERC20Metadata(tokens[i]).decimals();
```

Comment:
we could pass decimals as array. this works as well

Reply:
Yeah this reduces gas costs as well

Reply:
implemented